File: java/v5.0.2023/Prime.java
class divisor {
  public static int[] primes = new int[1000];
  public static int found = 4;
  public static int divisors;
}

class prime {
  public static int is_prime(int number) {
    int flag_prime = 1;
    for(int divider = 3; divider < (int) Math.sqrt(number) + 1; divider = divider + 2) {
      if(number % divider == 0) {
        flag_prime = 0;
        break;
      }
    }
    return flag_prime;
  }

  public static void find_primes(int limit) {
    divisor.primes[0] = 3;
    divisor.primes[1] = 5;
    divisor.primes[2] = 7; 
    divisor.divisors = 3;
    for(int number = 11; number <= limit; number = number + 2) {
      if(is_prime(number) > 0) {
        divisor.found++;
        divisor.primes[divisor.found - 2] =  number;
      }
    }
    divisor.divisors = divisor.found - 1;
  }

  public static int is_prime_fast(int number) {
    int largest_divider = (int)Math.sqrt(number);
    int flag_prime = 1;
    for(int i = 0; i < divisor.divisors; i++) {
      if(number % divisor.primes[i] == 0) {
        flag_prime = 0;
        break;
      }
      if(divisor.primes[i] > largest_divider)
      {
          break;
      }
    }
    return flag_prime;
  }

  public static void main(String[] args) {
    int last = 10000000;
    System.out.println("Calculation prime numbers to " + last + " in Java v5.0.2023.");
    long start = System.nanoTime();
    int limit = (int)Math.sqrt(last);
    if(limit % 2 == 0) {
      limit++;
    }
    System.out.print("Find primes numbers to " + limit + ". ");
    find_primes(limit);
    System.out.println("Found " + divisor.found + " primes. Now using them as divisors.");

    for(int number = limit + 2; number < last; number = number + 2) {
      divisor.found += is_prime_fast(number);
    }
    long end = System.nanoTime();
    System.out.print("I found " + divisor.found + " prime numbers. ");
    System.out.print("This took me " + ((end - start) / 1000000000.0) + " seconds.");
  }
}

File: python/v5.4.2024/prime.py
# prime v5.4.2024 with primes as a divider list up to 4294967295 = 2E32 - 1

import math, time, cpuinfo

def is_prime(number):
    global found
    flag_prime = 1
    for divider in range(3, int(math.sqrt(number)) + 1, 2):
        if number % divider == 0:
            flag_prime = 0
            break
    return flag_prime

def find_primes(largest):
    global primes
    global found
    for number in range(11, largest + 1, 2):
        if is_prime(number) > 0:
            found += 1
            primes.append(number)

def is_prime_fast(number):
    global found
    flag_prime = 1
    largest_divider = int(math.sqrt(number)) + 1
    for divider in primes:
        if number % divider == 0:
            flag_prime = 0
            break
        if divider > largest_divider:
            break
    return flag_prime

def elapsed_time(seconds):
    hours = int(seconds/3600)
    minutes = int(seconds/60 - hours*60)
    sec = int(seconds - minutes*60 - hours*3600)
    return(f"{hours}h {minutes}min {sec}s")

if __name__ == "__main__":
    last  = 2147483647     # 4294967295 is the limit for unsigned 32bit
    found = 4              # we start from 11, know 2, 3, 5, 7
    primes = [3, 5, 7]     # exclude 2 since we only test odd numbers
    print(f"Calculating prime numbers to {last} in Python with algorithm v5.4.2024")
    print(f"Running on a {cpuinfo.get_cpu_info()['brand_raw']}")
    start = time.perf_counter_ns()
    dot = start
    column = 0    
    largest_divider = int(math.sqrt(last))
    if largest_divider % 2 == 0:
        largest_divider += 1
    print(f'First find prime divisors up to {largest_divider}.')
    find_primes(largest_divider)
    print(f'Found {found} primes, now use them als divisors.')
    for number in range(largest_divider + 2, last, 2):
        found += is_prime_fast(number)
        if (time.perf_counter_ns() - dot) > 2000000000:
            print(".", end="")
            dot = time.perf_counter_ns()
            column += 1
            if column > 30:
                t = elapsed_time((time.perf_counter_ns() - start)/1000000000)
                print(f" {t} - {number} {int(number*100/last)}% ")
                column = 1
    if column > 0:
        print(" ")
    duration = (time.perf_counter_ns() - start)/1000000000
    print(f"This took: {duration:.9f} seconds. {elapsed_time(duration)}")
    print(f"I found {found} prime numbers. Should be 78498.")
